Upgrade Karaf to 4.4.7, Xtext/Xtend 2.37.0

holgerfriedrich opened the pull request:
This PR upgrades to the recently released Karaf version 4.4.7 🎉 and lifts Xtext/Xtend to 2.37.0.
Additionally, it removes workaround #4526 which should no longer be necessary.

This consists of 3 commits, PLEASE DO NOT SQUASH during merge.

Fixes #4283
Fixes various CVEs in BouncyCastle  1.77, Jetty, etc.

Upgrade Karaf from 4.4.6 to 4.4.7
* Sync runtime dependencies with Karaf 4.4.7, most notably:
   * PaxWeb 8.0.30
   * Jetty 9.4.56.v20240826
   * BouncyCastle 1.78.1
   * CXF 3.6.4
   * Jackson 2.18.2
   * JNA 5.16.0
   * JAXB 2.3.9
   * commons-io 2.17.0
   * commons-lang3 3.17.0
   * XBean 4.26
   * ASM 9.7.1
   * PaxLogging 2.2.8
 * Resolve itest runbundles

Upgrade Xtext/Xtend to 2.37.0
* Upgrade Xtext/Xtend from 2.36.0 to 2.37.0, see release notes:
  https://eclipse.dev/Xtext/releasenotes.html#/releasenotes/2024/11/19/version-2-37-0
  https://eclipse.dev/Xtext/xtend/releasenotes.html#/releasenotes/2024/11/19/version-2-37-0
* Upgrade dependencies
  * ecj from 3.36.0 to 3.39.0
  * gson from 2.10.1 to 2.11.0
  * classgraph to 4.8.176
  * guava from 3.33.0 to 3.33.1 

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").

ATTENTION: Don't use "git merge" when working with your pull request branch!
This can clutter your Git history and make your PR unusable.
Use "git rebase" instead. See this forum post for further details:
https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358

All PRs should be created using the "main" branch as base.
Important bugfixes are cherry-picked by maintainers to the patch release branch after a PR has been merged.

If your PR's code is not backward compatible with previous releases (which
should be avoided), add a message to the release notes by filing another PR:
https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst

# Title

Provide a short summary in the *Title* above. It will show up in the release notes.
For example:
- Introduce metadata for all add-ons
- Fix memory leak in ScriptedRuleProvider

# Description

Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the Core README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis?
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work

If your pull request contains a new contribution, it will likely take some time
before it is reviewed and processed by maintainers.
-->

andrewfg commented:
@holgerfriedrich did you see the comments on #3315 about our dependency on Jetty 9.x whereby Jetty 9.x is EOL but upgrading to Jetty 10 or 11 or 12 apparently requires Karaf 4.5 ??

So in other words your incremental Karaf 4.4.x upgrades are all fine and good, but do not resolve the issue that we are over dependendent on Jetty 9.x which is now THREE major versions out of date from current Jetty code.

J-N-K commented:
We can't do anything about that. Karaf 4.5 should bring at least Jetty 10, but TBH - looking at the current speed of development - I don't see that before next year.

andrewfg commented:
@J-N-K is there a hard point in our Karaf 4.4.x configuration that absolutely forbids importing Jetty 10+ ?? .. I am wondering if there may be some approach whereby we can try to decouple these inter dependencies. The current tight interdependence put us (IMHO) in a very risky situation if one or other of two code bases is eventually abandoned.

holgerfriedrich commented:
@andrewfg the last discussion about Jetty on the Karaf mailing list I have seen is there: https://lists.apache.org/thread/n30xb51gffwmfxs1ldw8zlvj21moszyh

I agree with @J-N-K, it is likely that a Jetty upgrade will not happen soon. 
I cannot tell what our options are (besides waiting).

Karaf 4.4.x upgrades are still useful. IIRC with the last upgrades we got ready for Java 21 and closed a relevant CVE in Jetty. 4.4.7 will get bouncy castle to a recent version and upgrade a few more dependencies.

J-N-K commented:
I wonder if Karaf is a dead-end. PaxWeb does not even support Jetty 12 (there is a version for Jetty 10 - PayWeb 9.x), but Karaf does not use that.

andrewfg commented:
I am still wondering of there is some way to build OH to import Jetty 12 directly into OH core, and exclude the older Jetty parts that would come indirectly via Karaf. That would give us much tighter control over versioning.

J-N-K commented:
The difficult thing is that we also use both, Jetty server and Jetty client. Since they have common components it would be a lot easier if we use the same version for both. And that's where the problem is: For the server part (especially the servlets) we depend on PaxWeb and that currently doesn't work with the newer versions of Jetty (11/12).

andrewfg commented:
Hmm.

It seems to me that we are in some jeopardy here..

- OH depends on Karaf
- OH depends (therefore) on Paxweb
- OH depends (therefore) on Jetty
- Current Paxweb (and therefore) Karaf use Jetty 9.x
- Jetty 9.x is EOL and THREE major versions behind the current 12.x
- Paxweb dev is resisting upgrading the Jetty dependency from 9.x to 12.x due to a) the amount of effort, and b) the risk of breaking changes 
- Karaf dev is (therefore) considering dropping Paxweb because (as mentioned above) it is not current
- Karaf dev is (therefore) considering switching from Jetty to (say) Tomcat

It seems to me that OH architecture council needs to do dome thinking. Do we accept our core to be so depend on above mentioned weak points in dependency support? And/or do we have a vision of the future path? 

Perhaps we need to contribute active PRs to Paxweb and (therefore) Karaf in order to bypass the bottleneck? And/or perhaps we need to make our own fork of Paxweb (and eventually Karaf)?

kaikreuzer commented:
> Paxweb dev is resisting upgrading the Jetty dependency from 9.x to 12.x due to a) the amount of effort, and b) the risk of breaking changes
> Karaf dev is (therefore) considering dropping Paxweb because (as mentioned above) it is not current
> Karaf dev is (therefore) considering switching from Jetty to (say) Tomcat

Do you have links to these discussions in the respective communities? First step might be to involve ourselves there.
A decision by Paxweb not to upgrade essentially means that they have given up the project...

J-N-K commented:
I don't think that is what @grgrzybek said about future support. The last discussion regarding Jetty 11 (as Pax Web 10) that I remember is https://lists.apache.org/thread/4lwbt9g0w707zomnq6cfx179n8w90d3h: It's not ready (and as far as I can see there is still no release for Pax Web 10), but there is work on that. I can't find any comment about not supporting Jetty 12.

Changing to Jetty 11 (or 12) is also a big step for us because we need to move from `javax` to `jakarta` namespace. We should also upgrade JAXB to version 4 then.

grgrzybek commented:
Yes - moving from javax to jakarta is a huge effort - that's why Pax Web 10 is still to be done...
And while WAR/WAB and even Whiteboard (which implies HttpService underneath) are actually ready, entire Pax Web 10 has other JavaEE deps to switch to Jakarta (like CDI or JSF) which is more problematic because CDI/JSF OSGi support is modest to say the least...

And because I feel that Pax Web 10 would be much less than Pax Web 8/9 (9 differs only by switching Jetty 9 to Jetty 10, which are still `javax.servlet`), I intentionally delay it until I have higher motivation (and more time).

And yes - Pax Web 10 is using Jetty 12 (but not in it's multi-servlet-api extent) - only `jakarta.servlet` package is supported.
Ideally Pax Web 10 should ship with Whiteboard tracker for `javax.servlet` based Whiteboard services that could be wrapped with `jakarta.servlet` API (servlets, filters, listeners, error pages, welcome files, JSPs, SCIs).

andrewfg commented:
> I don't think that is what @grgrzybek said about future support.

Hmm. You may be right. He certainly has added support for Jetty 12 at least on [this repo](
https://github.com/ops4j/org.ops4j.pax.web/pulls?q=is%3Apr+is%3Aclosed+jetty) .. but I am not sure if that is the branch on which Karaf 4.5 would (eventually) be based ??

J-N-K commented:
Pax Web 9 doesn't solve our issues with outdated Jetty (as Jetty 10 is also EoL for community support), though.

grgrzybek commented:
My work on future Pax Web 10 based on Tomcat 10 (soon - 11), Undertow 2.3 and Jetty 12 is here: https://github.com/ops4j/org.ops4j.pax.web/commits/main-jakarta/

And I am aware that if Karaf 4.5 needs JakartaEE 10 support for web-related OSGi CMPN specifications (https://docs.osgi.org/specification/osgi.cmpn/8.1.0/) then either Felix HTTP is needed as default implementation or work on Pax Web 10 has to be hastened...

Mind that OSGi CMPN 8.1 is quite relentless for web-related specifications:
* chapter 102: HttpService is gone
* chapter 128: [Web Applications specification is still JavaEE](https://docs.osgi.org/specification/osgi.cmpn/8.1.0/service.war.html)
* chapter 140: [Whiteboard specification is done and renamed](https://docs.osgi.org/specification/osgi.cmpn/8.1.0/service.servlet.html)

shikousa commented:
@grgrzybek Thanks for the information. It seems that for the Unifi binding in Openhab we need partitioned cookies. The current Pax Web 8/Jetty 9 does not support the setPartitioned() feature. Is there some kind of workaround available or do we need to wait for >= Pax Web 9/Jetty 10 in Karaf?

https://community.openhab.org/t/unifi-binding-beta-3-2-0-3-6-0/131156/180

grgrzybek commented:
Oh, I wasn't aware of _partitioned_ cookies...
Jetty added an attribute in https://github.com/jetty/jetty.project/commit/f82844e2a256c3426df029f028bf0fee8cd4879b
I think not all browsers (Safari?) support it and definitely it's not part of Servlet API.
So please don't expect a unified (Jetty / Tomcat / Undertow) Pax Web support anytime soon...

J-N-K commented:
> Oh, I wasn't aware of _partitioned_ cookies... Jetty added an attribute in [jetty/jetty.project@f82844e](https://github.com/jetty/jetty.project/commit/f82844e2a256c3426df029f028bf0fee8cd4879b) I think not all browsers (Safari?) support it and definitely it's not part of Servlet API. So please don't expect a unified (Jetty / Tomcat / Undertow) Pax Web support anytime soon...

I believe we need that on client side, but I vaguely remember that using Jetty 9 server and Jetty 10 client in the same OSGi container is difficult. I might be wrong on that.

grgrzybek commented:
> @grgrzybek I think partitioned cookies are already supported in Jetty 10. So isn't it possible that Pax Web 9 also supports it?

May I ask you please to create an issue here? https://github.com/ops4j/org.ops4j.pax.web/issues

I may look at it, as it sounds interesting and important ;) But I can't promise any release date with this implemented... Please provide some context and links there.

holgerfriedrich commented:
Short update:

Update to Xtext/Xtend 2.37.0 has been added. It relies on ASM 9.7.1, introduced with Karaf 4.4.7. Could not get the feature resolution to succeed on current master, this is why I added it here.

PRs for all repos are ready and up to date. Compilation and tests are running fine.

Once Karaf 4.4.7 is released, we are ready to go.

This is quite a risk as the planned winter release is soon and only one milestone build planned in between.
On the other hand, we should include the patch to Jetty 9.4.56 included in the release, if possible.

holgerfriedrich commented:
The issue above was caused by an outdated ref to 4.4.6 jar in karaf.bat.

The current karaf 4.4.7-SNAPSHOT built from 4.4.x branch seems to work fine.
OH starts up w/o warning (at default log level), wizard runs fine, add-ons install properly. My devices are added to the Inbox.

holgerfriedrich commented:
Reminder to remove workaround #4526

holgerfriedrich commented:
Good news, everybody!
We are close to the karaf-4.4.7 release, [vote for release](https://lists.apache.org/thread/g3n8xl3pcrpr416bcntxsp9zx0cjmt4h) has been started. 


---

I have installed the release version locally. OH build is fine, final tests with the release build are running....
My local distribution build seems to work. It starts up, and installation of plugins is fine as well.

Something odd came up, though. During startup
```
Launching the openHAB runtime...
WARNING: package org.apache.karaf.specs.locator not in java.base
Error starting karaf activator org.apache.karaf.specs.activator.Activator: org/apache/karaf/specs/locator/OsgiLocator
```

and later in the log:

```
20:38:41.398 [ERROR] [Events.Framework                     ] - FrameworkEvent ERROR
java.lang.NoClassDefFoundError: org/apache/karaf/specs/locator/OsgiLocator
        at org.apache.karaf.specs.activator.Activator.register(Activator.java:125) ~[org.apache.karaf.specs.activator-4.4.7.jar:4.4.7]
        at org.apache.karaf.specs.activator.Activator.bundleChanged(Activator.java:97) ~[org.apache.karaf.specs.activator-4.4.7.jar:4.4.7]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:949) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEventPrivileged(EquinoxEventPublisher.java:229) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:138) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:130) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor.publishModuleEvent(EquinoxContainerAdaptor.java:217) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.eclipse.osgi.container.ModuleContainer.applyDelta(ModuleContainer.java:857) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.eclipse.osgi.container.ModuleContainer.resolveAndApply(ModuleContainer.java:560) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.eclipse.osgi.container.ModuleContainer.resolve(ModuleContainer.java:503) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.eclipse.osgi.container.ModuleContainer.resolve(ModuleContainer.java:492) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.eclipse.osgi.container.ModuleContainer$ContainerWiring.resolveBundles(ModuleContainer.java:1451) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.apache.karaf.features.internal.service.BundleInstallSupportImpl.resolveBundles(BundleInstallSupportImpl.java:244) ~[?:?]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.resolveBundles(FeaturesServiceImpl.java:1175) ~[?:?]
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:1027) ~[?:?]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1069) ~[?:?]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:1004) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.lang.ClassNotFoundException: org.apache.karaf.specs.locator.OsgiLocator
        ... 23 more
```

I think I have not seen this in my build a few days ago.

holgerfriedrich commented:
Problem resolved, I overlooked to remove `-SNAPSHOT` in `karaf.bat` when I was switching to the final zip.
So, now everything looks good to me!
Big thanks to the Karaf team!

holgerfriedrich commented:
@wborn @kaikreuzer We are ready to go for the next Karaf upgrade. PRs are in place on all repos.

wborn commented:
Great! :+1: I'm give it a small test now. :slightly_smiling_face:
